utils/staking-miner/src/monitor.rs

```diff
@@ -0,0 +1,191 @@
+// Copyright 2021 Parity Technologies (UK) Ltd.
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! The monitor command.
+
+use crate::{prelude::*, rpc_helpers::*, params, Signer, SharedConfig, MonitorConfig, Error};
+use jsonrpsee_ws_client::{
+	traits::{SubscriptionClient},
+	v2::params::JsonRpcParams,
+	Subscription, WsClient,
+};
+
+/// Ensure that now is the singed phase.
+async fn ensure_signed_phase<T: EPM::Config, B: BlockT>(
+	client: &WsClient,
+	at: B::Hash,
+) -> Result<(), Error> {
+	let key = sp_core::storage::StorageKey(EPM::CurrentPhase::<T>::hashed_key().to_vec());
+	let phase = get_storage::<EPM::Phase<BlockNumber>>(client, params! {key, at}).await?.unwrap_or_default();
+
+	if phase.is_signed() {
+		Ok(())
+	} else {
+		Err(Error::IncorrectPhase)
+	}
+}
+
+/// Ensure that our current `us` have not submitted anything previously.
+async fn ensure_no_previous_solution<T: EPM::Config, B: BlockT>(
+	client: &WsClient,
+	at: B::Hash,
+	us: &AccountId,
+) -> Result<(), Error> {
+	use EPM::{SignedSubmissions, signed::SignedSubmission};
+	let key = sp_core::storage::StorageKey(SignedSubmissions::<T>::hashed_key().to_vec());
+	let queue = crate::any_runtime! {
+		get_storage::<Vec<SignedSubmission<AccountId, Balance, NposCompactSolution>>>(client, params!{ key, at }).await
+	}?.unwrap_or_default();
+
+	// if we have a solution in the queue, then don't do anything.
+	if queue.iter().any(|ss| &ss.who == us) {
+		Err(Error::AlreadySubmitted)
+	} else {
+		Ok(())
+	}
+}
+
+macro_rules! monitor_cmd_for { ($runtime:tt) => { paste::paste! {
+	/// The monitor command.
+	pub(crate) async fn [<monitor_cmd_ $runtime>](
+		client: WsClient,
+		shared: SharedConfig,
+		config: MonitorConfig,
+		signer: Signer, // TODO: Signer could also go into a shared config, perhaps.
+	) -> Result<(), Error> {
+		use $crate::[<$runtime _runtime_exports>]::*;
+		let subscription_method = if config.listen == "head" {
+			"chain_subscribeNewHeads"
+		} else {
+			"chain_subscribeFinalizedHeads"
+		};
+
+		log::info!(target: LOG_TARGET, "subscribing to {:?}", subscription_method);
+		let mut subscription: Subscription<Header> = client
+			.subscribe(&subscription_method, JsonRpcParams::NoParams, "unsubscribe")
+			.await
+			.unwrap();
+
+		loop {
+			let now = subscription.next().await.unwrap();
```

Comment:
```suggestion
		while let Some(now) = subscription.next().await {
```

Handle upstream connection closures more gracefully.